sm5/smartctl.8
\# Copyright (C) 2002-3 Bruce Allen <smartmontools-support@lists.sourceforge.net>
\#
\# $Id: smartctl.8,v 1.74 2003/07/19 10:21:37 ballen4705 Exp $
\# 
\# This program is free software; you can redistribute it and/or modify it
\# under the terms of the GNU General Public License as published by the Free
\# Software Foundation; either version 2, or (at your option) any later
\# version.
\# 
\# You should have received a copy of the GNU General Public License (for
\# example COPYING); if not, write to the Free Software Foundation, Inc., 675
\# Mass Ave, Cambridge, MA 02139, USA.
\#
\# This code was originally developed as a Senior Thesis by Michael Cornwell
\# at the Concurrent Systems Laboratory (now part of the Storage Systems
\# Research Center), Jack Baskin School of Engineering, University of
\# California, Santa Cruz. http://ssrc.soe.ucsc.edu/
\#
.TH SMARTCTL 8  "$Date: 2003/07/19 10:21:37 $" "smartmontools-5.1"
.SH NAME
smartctl \- Control and Monitor Utility for SMART Disks
.SH SYNOPSIS
.B smartctl [options] device

.SH DESCRIPTION
.B smartctl
controls the Self-Monitoring, Analysis and Reporting Technology
(SMART) system built into many ATA-3 and later ATA, IDE and
SCSI-3 hard drives. The purpose of SMART is to monitor the
reliability of the hard drive and predict drive failures, and to carry
out different types of drive self-tests.  This version of \fB smartctl\fP is
compatible with ATA/ATAPI-5 and earlier standards (see REFERENCES
below)

.B smartctl
is a command line utility designed to perform SMART tasks such as
printing the SMART self-test and error logs, and enabling and
disabling SMART automatic testing. Note: if the user issues a
SMART command that is (apparently) not implemented by the device,
.B smartctl
will print a warning message but issue the command anyway.  This should
not cause problems: unimplemented SMART commands issued to a
drive are ignored and return an error.

.B smartctl
also provides support for polling TapeAlert messages from SCSI tape drives and changers.

The user must specify the device to be controlled or interrogated as the final
argument to
.B smartctl.
ATA devices use the form "/dev/hd*" and SCSI devices use the form "/dev/sd*".
For SCSI Tape Drives and Changers with TapeAlert support use the devices
"/dev/nst*" and "/dev/sg*".  More general paths may also be specified.
.B smartctl
will attempt to guess the device type, but the '\-d' option can be used to
specify a device type of ATA or SCSI if required.

Note that the printed output of
.B smartctl
displays most numerical values in base 10 (decimal), but some values
are displayed in base 16 (hexidecimal).  To distinguish them, the base
16 values are always displayed with a leading
.B "0x"
for example: "0xff". This man page follows the same convention.

.PP
.SH OPTIONS
.PP
The options are grouped below into several categories.
.B smartctl
will execute the corresponding commands in the order: INFORMATION,
ENABLE/DISABLE, DISPLAY DATA, RUN/ABORT TESTS.

SCSI devices only accept the options 
.B \-h, \-?, \-V, \-i, \-a, \-d, \-s, \-H, \-t, \-C, \-l selftest, \-l error
and 
.B \-X. 
TapeAlert devices only accept the options 
.B \-h, \-?, \-V, \-i, \-a, \-d, \-s, \-t, \-l selftest, \-l error
and 
.B \-H.

Long options  are  not  supported  on  all  systems.   Use
.B 'smartctl \-h'
to see the available options.

.TP
.B SHOW INFORMATION:
.TP
.B \-h, \-\-help, \-\-usage
Prints a usage message to STDOUT and exits.
.TP
.B \-?
Same as
.B \-h.
.TP
.B \-V, \-\-version, \-\-copyright, \-\-license
Prints version, copyright, license, home page and CVS-id information for your
copy of
.B smartctl
to STDOUT and then exits.  Please include this information if you are
reporting bugs or problems.
.TP
.B \-i, \-\-info
Prints the device model number, serial number, firmware version, and ATA Standard
version/revision information.  Says if the device supports SMART, and if
so, whether SMART support is currently enabled or disabled.
.TP
.B \-a, \-\-all
Prints all SMART information about the disk or TapeAlert information
about the tape drive or changer.
This is equivalent to '\-H
\-i \-c \-A \-l error \-l selftest' (for SCSI, '\-H \-i \-l error
\-l selftest').
.TP
.B RUN-TIME BEHAVIOR:
.TP
.B \-q TYPE, \-\-quietmode=TYPE
Specifies that
.B smartctl
should run in one of the two quiet modes described here.  The valid arguments
to this option are:

.I errorsonly
\- only print: For the '\-l error' option, if nonzero, the number
of errors recorded in the SMART error log and the power-on time when
they occurred; For the '\-l selftest' option, errors recorded in the device
self-test log; For the '\-H' option, SMART "disk failing" status or device
Attributes (pre-failure or usage) which failed either now or in the
past; For the '\-A' option, device Attributes (pre-failure or usage)
which failed either now or in the past.

.I silent
\- print no output.  The only way to learn about what was
found is to use the exit status of
.B smartctl
(see RETURN VALUES below).
.TP
.B \-d TYPE, \-\-device=TYPE
Specifies the type of the device.  The valid arguments to this option
are \fIata\fP and \fIscsi\fP.  If this option is not used then
.B smartctl
will attempt to guess the device type from the device name.
.TP
.B \-T TYPE, \-\-tolerance=TYPE
Specifies how tolerant
.B smartctl
should be of SMART command failures.  The valid arguments to this option
are:

.I normal
\- exit on failure of a mandatory SMART command, but not on failure of an
optional SMART command.  This is the default.  

.I conservative
\- exit on failure of any SMART command.

.I permissive
\- ignore failure of any SMART command.

Here "mandatory" means "required by the ATA/ATAPI-5 Specification if the
device implements the SMART command set" and "optional" means "not
required by the ATA/ATAPI-5 Specification even if the device implements
the SMART command set."  The 'mandatory' SMART commands are: (1)
Enable/Disable Attribute Autosave, (2) Enable/Disable SMART, and (3)
SMART Return Status.

.TP
.B \-b TYPE, \-\-badsum=TYPE
Specifies the action 
.B smartctl
should take if a checksum error is detected in the: (1) Device
Identity Structure, (2) SMART Self-Test Log Structure, (3)
SMART Attribute Value Structure, (4) SMART Attribute
Threshold Structure, or (5) ATA Error Log Structure.

The valid arguments to this option are:

.I warn
\- report the incorrect checksum but carry on in spite of it.  This is the
default.

.I exit
\- exit
.B smartctl.

.I ignore
\- continue silently without issuing a warning.

.TP
.B \-r TYPE, \-\-report=TYPE
Intended primarily to help
.B smartmontools
developers understand the behavior of
.B smartmontools
on non-conforming or poorly conforming hardware.  This option reports
details of
.B smartctl
transactions with the device.  The option can be used multiple times.

sm5/NEWS
smartmontools NEWS
------------------
CVS ID: $Id: NEWS,v 1.25 2004/09/14 01:48:35 ballen4705 Exp $

The most up-to-date version of this file is:
http://cvs.sourceforge.net/viewcvs.py/smartmontools/sm5/NEWS?sortby=date&view=markup


Date 2004-9-5
Summary: smartmontools release 5.33 (UNSTABLE/EXPERIMENTAL)
-----------------------------------------------------------
This is an unstable/experimental release of smartmontools.  It includes
 - support for Darwin/Mac OSX
 - support for OpenBSD
 - support for 3ware ATA RAID controllers under FreeBSD
 - support for 3ware 9500 series SATA RAID controllers under
   Linux.  Use /dev/twa[0-15] devices to access these.
 - support for 3ware character device interfaces /dev/twe[0-15]
   under Linux.  This allows (for example) Selective Self-tests.
 - support for Marvell chip-set based SATA controllers under Linux.
Please report sucess/failure with these items to the
smartmontools-support mailing list.


Date 2004-7-5
Summary: smartmontools release 5.32 (STABLE)
-----------------------------------------------------------
This is an stable release of smartmontools.
Note added 2004/7/7: users building a Solaris/Intel version of the code should 
modify the 'configure' file, changing "pc-*-solaris*" on line 106
to read "*-pc-solaris*".  Reference:
http://cvs.sourceforge.net/viewcvs.py/smartmontools/sm5/configure.in?r1=1.83&r2=1.84


Date: 2004-5-4
Summary: smartmontools release 5.31 (UNSTABLE/EXPERIMENTAL)
-----------------------------------------------------------
This is an unstable/experimental release of smartmontools.  It includes
several new features:
- Windows smartd daemon
- smartd now monitors current and pending sector counts
- Support for ATA-7 selective self-test features (Linux/NetBSD only)
  Please report sucess/failure with this option to the smartmontools-support
  mailing list.

Date: 2004-3-6
Summary: smartmontools release 5.30 (STABLE)
--------------------------------------------
This is a stable release of smartmontools: the first stable release
since 5.26.
- KNOWN BUG (identified/fixed by CF): smartd will segv and crash if
  the configuration file /etc/smartd.conf contains NO valid entries.
  This bug was introduced in version 1.259 of smartd.c by BA and
  is present in smartmontools releases 5.27-5.30 inclusive. This can
  be fixed by editing line 3165 of smartd.c, and changing:
  "else if (cfgentries[0]) {"
  to read:
  "else if (cfgentries && cfgentries[0]) {"


Date: 2004-2-24
Summary: smartmontools release 5.29 (Experimental, not STABLE)
--------------------------------------------------------------
This is another experimental release, to replace the 5.27 release that
had a damaged configure script.  The next stable release will be 5.30
- This release has SCSI support for NetBSD


Date: 2004-2-12 
Summary: smartmontools release 5.27 (Experimental, not STABLE)
--------------------------------------------------------------
- WARNING: this release has a broken --prefix=/a/path option to the
  configure script.  The consequence is that smartd will not look for the
  configuration file (smartd.conf) at the desired location.
- NetBSD support added
- A new Directive (-s) for smartd.conf now enables flexible automatic
  scheduled self-testing for both ATA and SCSI devices.
- Solaris now has ATA device support (SPARC only)
- A new Directive (-n) for smartd.conf to avoid spinning up disks

sm5/smartctl.c
/*
 * smartctl.c
 *
 * Home page of code is: http://smartmontools.sourceforge.net
 *
 * Copyright (C) 2002-3 Bruce Allen <smartmontools-support@lists.sourceforge.net>
 * Copyright (C) 2000 Michael Cornwell <cornwell@acm.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * You should have received a copy of the GNU General Public License
 * (for example COPYING); if not, write to the Free
 * Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * This code was originally developed as a Senior Thesis by Michael Cornwell
 * at the Concurrent Systems Laboratory (now part of the Storage Systems
 * Research Center), Jack Baskin School of Engineering, University of
 * California, Santa Cruz. http://ssrc.soe.ucsc.edu/
 *
 */


#include <stdio.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <linux/hdreg.h>
#include <sys/fcntl.h>
#include <sys/types.h>
#include <string.h>
#include <stdarg.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#include "atacmds.h"
#include "ataprint.h"
#include "extern.h"
#include "knowndrives.h"
#include "scsicmds.h"
#include "scsiprint.h"
#include "smartctl.h"
#include "utility.h"

extern const char *atacmds_c_cvsid, *ataprint_c_cvsid, *knowndrives_c_cvsid, *scsicmds_c_cvsid, *scsiprint_c_cvsid, *utility_c_cvsid; 
const char* smartctl_c_cvsid="$Id: smartctl.c,v 1.68 2003/04/16 01:54:39 dpgilbert Exp $"
ATACMDS_H_CVSID ATAPRINT_H_CVSID EXTERN_H_CVSID KNOWNDRIVES_H_CVSID SCSICMDS_H_CVSID SCSIPRINT_H_CVSID SMARTCTL_H_CVSID UTILITY_H_CVSID;

// This is a block containing all the "control variables".  We declare
// this globally in this file, and externally in other files.
smartmonctrl *con=NULL;


void printslogan(){
  pout("smartctl version %d.%d-%d Copyright (C) 2002-3 Bruce Allen\n",
      (int)RELEASE_MAJOR, (int)RELEASE_MINOR, (int)SMARTMONTOOLS_VERSION);
  pout("Home page is %s\n\n",PROJECTHOME);
  return;
}


void printcopy(){
  char out[CVSMAXLEN];
  pout("smartctl comes with ABSOLUTELY NO WARRANTY. This\n");
  pout("is free software, and you are welcome to redistribute it\n");
  pout("under the terms of the GNU General Public License Version 2.\n");
  pout("See http://www.gnu.org for further details.\n\n");
  pout("CVS version IDs of files used to build this code are:\n");
  printone(out,atacmds_c_cvsid);
  pout("%s",out);
  printone(out,ataprint_c_cvsid);
  pout("%s",out);
  printone(out,knowndrives_c_cvsid);
  pout("%s",out);
  printone(out,scsicmds_c_cvsid);
  pout("%s",out);
  printone(out,scsiprint_c_cvsid);
  pout("%s",out);
  printone(out,smartctl_c_cvsid);
  pout("%s",out);
  printone(out,utility_c_cvsid);
  pout("%s",out);
  return;
}

/*  void prints help information for command syntax */
void Usage (void){
  printf("Usage: smartctl [options] [device]\n");
  printf("\n==============  SHOW INFORMATION OPTIONS  =================================\n");
#ifdef HAVE_GETOPT_LONG
  printf(
"  -h, -?, --help, --usage\n"
"         Display this help and exit\n"
"  -V, --version, --copyright, --license\n"
"         Print license, copyright, and version information and exit\n"
"  -i, --info                                                       \n"
"         Show identity information for device\n"
"  -a, --all                                                        \n"
"         Show all SMART information for device\n"
  );
#else
  printf(
"  -h, -?    Display this help and exit\n"
"  -V        Print license, copyright, and version information\n"
"  -i        Show identity information for device\n"
"  -a        Show all SMART information for device                   \n"
  );
#endif
  printf("==============  SMARTCTL RUN-TIME BEHAVIOR OPTIONS  =======================\n");
#ifdef HAVE_GETOPT_LONG
  printf(
"  -q TYPE, --quietmode=TYPE                                           (ATA)\n"
"         Set smartctl quiet mode to one of: errorsonly, silent\n"
"  -d TYPE, --device=TYPE\n"
"         Specify device type to one of: ata, scsi\n"
"  -T TYPE, --tolerance=TYPE                                           (ATA)\n"
"         Set tolerance to one of: normal, conservative, permissive\n"
"  -b TYPE, --badsum=TYPE                                              (ATA)\n"
"         Set action on bad checksum to one of: warn, exit, ignore\n"
"  -r TYPE, --report=TYPE\n"
"         Report transactions for one of: ioctl, ataioctl, scsiioctl\n"
  );
#else
  printf(
"  -q TYPE   Set smartctl quiet mode to one of: errorsonly, silent     (ATA)\n"
"  -d TYPE   Specify device type to one of: ata, scsi\n"
"  -T TYPE   Set tolerance to one of: normal, conservative, permissive (ATA)\n"
"  -b TYPE   Set action on bad checksum to one of: warn, exit, ignore  (ATA)\n"
"  -r TYPE   Report transactions for one of: ioctl, ataioctl, scsiioctl\n"
  );
#endif
  printf("==============  DEVICE FEATURE ENABLE/DISABLE COMMANDS  ===================\n");
#ifdef HAVE_GETOPT_LONG
  printf(
"  -s VALUE, --smart=VALUE\n"
"        Enable/disable SMART on device (on/off)\n"
"  -o VALUE, --offlineauto=VALUE                                       (ATA)\n"
"        Enable/disable automatic offline testing on device (on/off)\n"
"  -S VALUE, --saveauto=VALUE                                          (ATA)\n"
"        Enable/disable Attribute autosave on device (on/off)\n"
  );
#else
  printf(
"  -s VALUE  Enable/disable SMART on device (on/off)\n"
"  -o VALUE  Enable/disable device automatic offline testing (on/off)  (ATA)\n"
"  -S VALUE  Enable/disable device Attribute autosave (on/off)         (ATA)\n"
  );
#endif
  printf("==============  READ AND DISPLAY DATA OPTIONS  ============================\n");
#ifdef HAVE_GETOPT_LONG
  printf(
"  -H, --health\n"
"        Show device SMART health status\n"
"  -c, --capabilities                                                  (ATA)\n"
"        Show device SMART capabilities\n"
"  -A, --attributes                                                    (ATA)\n"
"        Show device SMART vendor-specific Attributes and values\n"
"  -l TYPE, --log=TYPE\n"
"        Show device log. Type is one of: error, selftest\n"
"  -v N,OPTION , --vendorattribute=N,OPTION                            (ATA)\n"
"        Set display OPTION for vendor Attribute N (see man page)\n"
"  -F, --fixbyteorder                                                  (ATA)\n"
"        Fix byte order in some SMART data (some Samsung disks)\n"
"  -P TYPE, --presets=TYPE                                             (ATA)\n"
"        Drive-specific presets: use, ignore, show, showall\n"
  );
#else
  printf(
"  -H        Show device SMART health status\n"
"  -c        Show device SMART capabilities                            (ATA)\n"
"  -A        Show device SMART vendor-specific Attributes and values   (ATA)\n"
"  -l TYPE   Show device log. Type is one of: error, selftest\n"
"  -v N,OPT  Set display OPTion for vendor Attribute N (see man page)  (ATA)\n"
"  -F        Fix byte order in some SMART data (some Samsung disks)    (ATA)\n"
"  -P TYPE   Drive-specific presets: use, ignore, show, showall        (ATA)\n"
  );
#endif
  printf("==============  DEVICE SELF-TEST OPTIONS  =================================\n");
#ifdef HAVE_GETOPT_LONG
  printf(
"  -t TEST, --test=TEST\n"
"        Run test on device.  TEST is one of: offline, short, long\n"
"  -C, --captive\n"
"        With -t, performs test in captive mode (short/long only)\n"
"  -X, --abort\n"
"        Abort any non-captive test on device\n"
);
#else
  printf(
"  -t TEST   Run test on device.  TEST is one of: offline, short, long   \n"
"  -C        With -t, performs test in captive mode (short/long only)  \n"
"  -X        Abort any non-captive test                                \n"
  );
#endif
  printf("==============  SMARTCTL EXAMPLES  ========================================\n");
#ifdef HAVE_GETOPT_LONG
  printf(
"  smartctl -a /dev/hda                 (Prints all SMART information)\n"
"  smartctl --smart=on --offlineauto=on --saveauto=on /dev/hda\n"
"                                       (Enables SMART on first disk)\n"
"  smartctl -t long /dev/hda            (Executes extended disk self-test)\n"
"  smartctl --attributes --log=selftest --quietmode=errorsonly /dev/hda\n"
"                                       (Prints Self-Test & Attribute errors)\n"
  );
#else
  printf(
"  smartctl -a /dev/hda                 (Prints all SMART information)\n"
"  smartctl -s on -o on -S on /dev/hda  (Enables SMART on first disk)\n"
"  smartctl -t long /dev/hda            (Executes extended disk self-test)\n"
"  smartctl -A -l selftest -q errorsonly /dev/hda\n"
"                                       (Prints Self-Test & Attribute errors)\n"
  );
#endif
}

/* Returns a pointer to a static string containing a formatted list of the valid
   arguments to the option opt or NULL on failure. */
const char *getvalidarglist(char opt) {
  static char *v_list = NULL;
  char *s;

  switch (opt) {
  case 'q':
    return "errorsonly, silent";
  case 'd':
    return "ata, scsi";
  case 'T':
    return "normal, conservative, permissive";
  case 'b':
    return "warn, exit, ignore";
  case 'r':
    return "ioctl, ataioctl, scsiioctl";
  case 's':
  case 'o':
  case 'S':
    return "on, off";
  case 'l':
    return "error, selftest";
  case 'v':
    if (v_list) 
      return v_list;
    if (!(s = create_vendor_attribute_arg_list()))
      return NULL;
    // Allocate space for tab + "help" + newline + s + terminating 0
    v_list = (char *)malloc(7+strlen(s));
    sprintf(v_list, "\thelp\n%s", s);
    free(s);
    return v_list;
  case 'P':
    return "use, ignore, show, showall";
  case 't':
    return "offline, short, long";
  default:
    return NULL;
  }
}

/* Prints the message "=======> VALID ARGUMENTS ARE: <LIST>  <=======\n", where
   <LIST> is the list of valid arguments for option opt. */
void printvalidarglistmessage(char opt) {
  const char *s;
  char separator;

  if (!(s = getvalidarglist(opt))) {
    pout("Error whilst constructing argument list for option %c", opt);
    return;
  }

  // getvalidarglist() might produce a multiline or single line string.  We
  // need to figure out which to get the formatting right.
  separator = strchr(s, '\n') ? '\n' : ' ';

  pout("=======> VALID ARGUMENTS ARE:%c%s%c<=======\n", separator, (char *)s,
    separator);
}

unsigned char tryata=0,tryscsi=0;

/*      Takes command options and sets features to be run */	
void ParseOpts (int argc, char** argv){
  int optchar;
  int badarg;
  int captive;
  extern char *optarg;
  extern int optopt, optind, opterr;
  // Please update getvalidarglist() if you edit shortopts
  const char *shortopts = "h?Vq:d:T:b:r:s:o:S:HcAl:iav:P:t:CXF";
#ifdef HAVE_GETOPT_LONG
  char *arg;
  // Please update getvalidarglist() if you edit longopts
  struct option longopts[] = {
    { "help",            no_argument,       0, 'h' },
    { "usage",           no_argument,       0, 'h' },
    { "version",         no_argument,       0, 'V' },
    { "copyright",       no_argument,       0, 'V' },
    { "license",         no_argument,       0, 'V' },
    { "quietmode",       required_argument, 0, 'q' },
    { "device",          required_argument, 0, 'd' },
    { "tolerance",       required_argument, 0, 'T' },
    { "badsum",          required_argument, 0, 'b' },
    { "report",          required_argument, 0, 'r' },
    { "smart",           required_argument, 0, 's' },
    { "offlineauto",     required_argument, 0, 'o' },
    { "saveauto",        required_argument, 0, 'S' },
    { "health",          no_argument,       0, 'H' },
    { "capabilities",    no_argument,       0, 'c' },
    { "attributes",      no_argument,       0, 'A' },
    { "log",             required_argument, 0, 'l' },
    { "info",            no_argument,       0, 'i' },
    { "all",             no_argument,       0, 'a' },
    { "vendorattribute", required_argument, 0, 'v' },
    { "presets",         required_argument, 0, 'P' },
    { "test",            required_argument, 0, 't' },
    { "captive",         no_argument,       0, 'C' },
    { "abort",           no_argument,       0, 'X' },
    { "fixbyteorder",    no_argument,       0, 'F' },
    { 0,                 0,                 0, 0   }
  };
#endif
  
  memset(con,0,sizeof(*con));
  con->testcase=-1;
  opterr=optopt=0;
  badarg = captive = FALSE;
  
  // This miserable construction is needed to get emacs to do proper indenting. Sorry!
  while (-1 != (optchar = 
#ifdef HAVE_GETOPT_LONG
		getopt_long(argc, argv, shortopts, longopts, NULL)
#else
		getopt(argc, argv, shortopts)
#endif
		)){
    switch (optchar){
    case 'V':
      con->veryquietmode=FALSE;
      printslogan();
      printcopy();
      exit(0);
      break;
    case 'q':
      if (!strcmp(optarg,"errorsonly")) {
        con->quietmode     = TRUE;
        con->veryquietmode = FALSE;
      } else if (!strcmp(optarg,"silent")) {
        con->veryquietmode = TRUE;
        con->quietmode     = TRUE;
      } else {
        badarg = TRUE;
      }
      break;
    case 'd':
      if (!strcmp(optarg,"ata")) {
        tryata  = TRUE;
        tryscsi = FALSE;
      } else if (!strcmp(optarg,"scsi")) {
        tryata  = FALSE;
        tryscsi = TRUE;
      } else {
        badarg = TRUE;
      }
      break;
    case 'T':
      if (!strcmp(optarg,"normal")) {
        con->conservative = FALSE;
        con->permissive   = FALSE;
      } else if (!strcmp(optarg,"conservative")) {
        con->conservative = TRUE;
        con->permissive   = FALSE;
      } else if (!strcmp(optarg,"permissive")) {
        con->permissive   = TRUE;
        con->conservative = FALSE;
      } else {
        badarg = TRUE;
      }
      break;
    case 'b':
      if (!strcmp(optarg,"warn")) {
        con->checksumfail   = FALSE;
        con->checksumignore = FALSE;
      } else if (!strcmp(optarg,"exit")) {
        con->checksumfail   = TRUE;
        con->checksumignore = FALSE;
      } else if (!strcmp(optarg,"ignore")) {
        con->checksumignore = TRUE;
        con->checksumfail   = FALSE;
      } else {
        badarg = TRUE;
      }
      break;
    case 'r':
      if (!strcmp(optarg,"ioctl")) {
        con->reportataioctl++;
        con->reportscsiioctl++;
      } else if (!strcmp(optarg,"ataioctl")) {
        con->reportataioctl++;
      } else if (!strcmp(optarg,"scsiioctl")) {
        con->reportscsiioctl++;
      } else {
        badarg = TRUE;
      }
      break;
    case 's':
      if (!strcmp(optarg,"on")) {
        con->smartenable  = TRUE;
        con->smartdisable = FALSE;
      } else if (!strcmp(optarg,"off")) {
        con->smartdisable = TRUE;
        con->smartenable  = FALSE;
      } else {
        badarg = TRUE;
      }
      break;
    case 'o':
      if (!strcmp(optarg,"on")) {
        con->smartautoofflineenable  = TRUE;
        con->smartautoofflinedisable = FALSE;
      } else if (!strcmp(optarg,"off")) {
        con->smartautoofflinedisable = TRUE;
        con->smartautoofflineenable  = FALSE;
      } else {
        badarg = TRUE;
      }
      break;
    case 'S':
      if (!strcmp(optarg,"on")) {
        con->smartautosaveenable  = TRUE;
        con->smartautosavedisable = FALSE;
      } else if (!strcmp(optarg,"off")) {
        con->smartautosavedisable = TRUE;
        con->smartautosaveenable  = FALSE;
      } else {
        badarg = TRUE;
      }
      break;
    case 'H':
      con->checksmart = TRUE;		
      break;
    case 'F':
      con->reversesamsung = TRUE;
      break;
    case 'c':
      con->generalsmartvalues = TRUE;
      break;
    case 'A':
      con->smartvendorattrib = TRUE;
      break;
    case 'l':
      if (!strcmp(optarg,"error")) {
        con->smarterrorlog = TRUE;
      } else if (!strcmp(optarg,"selftest")) {
        con->smartselftestlog = TRUE;
      } else if (!strcmp(optarg,"directory")) {
        con->smartlogdirectory = TRUE;
      } else {
        badarg = TRUE;
      }
      break;
    case 'i':
      con->driveinfo = TRUE;
      break;		
    case 'a':
      con->driveinfo          = TRUE;
      con->checksmart         = TRUE;
      con->generalsmartvalues = TRUE;
      con->smartvendorattrib  = TRUE;
      con->smarterrorlog      = TRUE;
      con->smartselftestlog   = TRUE;
      break;
    case 'v':
      // parse vendor-specific definitions of attributes
      if (!strcmp(optarg,"help")) {
        char *s;
        con->veryquietmode=FALSE;
        printslogan();
        if (!(s = create_vendor_attribute_arg_list())) {
          pout("Insufficient memory to construct argument list\n");
          exit(FAILCMD);
        }
        pout("The valid arguments to -v are:\n\thelp\n%s\n", s);
        free(s);
        exit(0);
      }
      if (parse_attribute_def(optarg, con->attributedefs))
	badarg = TRUE;
      break;    
    case 'P':
      if (!strcmp(optarg, "use")) {
        con->ignorepresets = FALSE;
      } else if (!strcmp(optarg, "ignore")) {
        con->ignorepresets = TRUE;
      } else if (!strcmp(optarg, "show")) {
        con->showpresets = TRUE;
      } else if (!strcmp(optarg, "showall")) {
        showallpresets();
        exit(0);
      } else {
        badarg = TRUE;
      }
      break;
    case 't':
      if (!strcmp(optarg,"offline")) {
        con->smartexeoffimmediate = TRUE;
        con->testcase             = OFFLINE_FULL_SCAN;
      } else if (!strcmp(optarg,"short")) {
        con->smartshortselftest = TRUE;
        con->testcase           = SHORT_SELF_TEST;
      } else if (!strcmp(optarg,"long")) {
        con->smartextendselftest = TRUE;
        con->testcase            = EXTEND_SELF_TEST;
      } else {
        badarg = TRUE;
      }
      break;
    case 'C':
      captive = TRUE;
      break;
    case 'X':
      con->smartselftestabort = TRUE;
      con->testcase           = ABORT_SELF_TEST;
      break;
    case 'h':
    case '?':
    default:
      con->veryquietmode=FALSE;
      printslogan();
#ifdef HAVE_GETOPT_LONG
      // Point arg to the argument in which this option was found.
      arg = argv[optind-1];
      // Check whether the option is a long option that doesn't map to -h.
      if (arg[1] == '-' && optchar != 'h') {
        // Iff optopt holds a valid option then argument must be missing.
        if (optopt && (strchr(shortopts, optopt) != NULL)) {
          pout("=======> ARGUMENT REQUIRED FOR OPTION: %s <=======\n", arg+2);
          printvalidarglistmessage(optopt);
	} else
	  pout("=======> UNRECOGNIZED OPTION: %s <=======\n",arg+2);
	pout("\nUse smartctl --help to get a usage summary\n\n");
	exit(FAILCMD);
      }
#endif
      if (optopt) {
        // Iff optopt holds a valid option then argument must be missing.
        if (strchr(shortopts, optopt) != NULL) {
          pout("=======> ARGUMENT REQUIRED FOR OPTION: %c <=======\n", optopt);
          printvalidarglistmessage(optopt);
        } else
	  pout("=======> UNRECOGNIZED OPTION: %c <=======\n",optopt);
	pout("\nUse smartctl -h to get a usage summary\n\n");
	exit(FAILCMD);
      }
      Usage();
      exit(0);	
    } // closes switch statement to process command-line options
    
    // Check to see if option had an unrecognized or incorrect argument.
    if (badarg) {
      printslogan();
      // It would be nice to print the actual option name given by the user
      // here, but we just print the short form.  Please fix this if you know
      // a clean way to do it.
      pout("=======> INVALID ARGUMENT TO -%c: %s <======= \n", optchar, optarg);
      printvalidarglistmessage(optchar);
      pout("\nUse smartctl -h to get a usage summary\n\n");
      exit(FAILCMD);
    }
  }
  // At this point we have processed all command-line options.

  // Do this here, so results are independent of argument order	
  if (con->quietmode)
    con->veryquietmode=TRUE;
  
  // error message if user has asked for more than one test
  if (1<(con->smartexeoffimmediate+con->smartshortselftest+con->smartextendselftest+
	 con->smartshortcapselftest+con->smartextendcapselftest+con->smartselftestabort)){
    con->veryquietmode=FALSE;
    printslogan();
    pout("\nERROR: smartctl can only run a single test (or abort) at a time.\n");
    pout("Use smartctl -h to get a usage summary\n\n");
    exit(FAILCMD);
  }

  // If captive option was used, change test type if appropriate.
  if (captive && con->smartshortselftest) {
      con->smartshortselftest    = FALSE;
      con->smartshortcapselftest = TRUE;
      con->testcase              = SHORT_CAPTIVE_SELF_TEST;
  } else if (captive && con->smartextendselftest) {
      con->smartextendselftest    = FALSE;
      con->smartextendcapselftest = TRUE;
      con->testcase               = EXTEND_CAPTIVE_SELF_TEST;
  }

  // From here on, normal operations...
  printslogan();
  
  // Warn if the user has provided no device name
  if (argc-optind<1){
    pout("ERROR: smartctl requires a device name as the final command-line argument.\n\n");
    pout("Use smartctl -h to get a usage summary\n\n");
    exit(FAILCMD);
  }
  
  // Warn if the user has provided more than one device name
  if (argc-optind>1){
    int i;
    pout("ERROR: smartctl takes ONE device name as the final command-line argument.\n");
    pout("You have provided %d device names:\n",argc-optind);
    for (i=0; i<argc-optind; i++)
      pout("%s\n",argv[optind+i]);
    pout("Use smartctl -h to get a usage summary\n\n");
    exit(FAILCMD);
  }  
}

// Printing function (controlled by global con->veryquietmode) 
// [From GLIBC Manual: Since the prototype doesn't specify types for
// optional arguments, in a call to a variadic function the default
// argument promotions are performed on the optional argument
// values. This means the objects of type char or short int (whether
// signed or not) are promoted to either int or unsigned int, as
// appropriate.]
void pout(char *fmt, ...){
  va_list ap;
  
  // initialize variable argument list 
  va_start(ap,fmt);
  if (con->veryquietmode){
    va_end(ap);
    return;
  }

  // print out
  vprintf(fmt,ap);
  va_end(ap);
  fflush(stdout);
  return;
}


/* Main Program */
int main (int argc, char **argv){
  int fd,retval=0;
  char *device;
  smartmonctrl control;
  const char *devroot="/dev/";

  // define control block for external functions
  con=&control;

  // Part input arguments
  ParseOpts(argc,argv);

  // open device - read-write access is needed to use the scsi generic (sg)
  // device for some commands (e.g. '-s on'). Attempt to open read-write and 
  // if that fails with an error suggesting writing is not permitted, fall
  // back to a read-only open. [When opened O_RDONLY invoking '-s on' will 
  // attempt the scsi command MODE SELECT. Sg disallows it since that scsi 
  // command potentially modifies mode page data.]
  fd = open(device=argv[argc-1], O_RDWR);
  if ((fd < 0) && ((EACCES == errno) || (EROFS == errno)))
    fd = open(device=argv[argc-1], O_RDONLY);  
  if (fd<0) {
    char errmsg[256];
    snprintf(errmsg,256,"Smartctl open device: %s failed",argv[argc-1]);
    errmsg[255]='\0';
    syserror(errmsg);
    return FAILDEV;
  }

  // if necessary, try to guess if this is an ATA or SCSI device
  if (!tryata && !tryscsi) {
    if (!strncmp(device,devroot,strlen(devroot)) && strlen(device)>5){
      if (device[5] == 'h')
	tryata=1;
      if (device[5] == 's')
	tryscsi=1;
    }
    else if (strlen(device)){
	if (device[0] == 'h')
	  tryata=1;
	if (device[0] == 's')
	  tryscsi=1;
      }
  }

  // now call appropriate ATA or SCSI routine
  if (tryata)
    retval=ataPrintMain(fd);
  else if (tryscsi)
    scsiPrintMain (device, fd);
  else {
    pout("Smartctl: specify if this is an ATA or SCSI device with the -d option.\n");
    Usage();
    return FAILCMD;
  }

  return retval;
}
